### ISSUE TITLE
Some hints are incorrect with CPP

### ISSUE THREAD

**Issue by zliu41** [ndmitchell/hlint#947]
HLint suggests eta reduce for:
```haskell
foo x = bar y x
#if defined(__GLASGOW_HASKELL__)
  where y = x
#endif
```

Same problem with all "Redundant X" hints.

Maybe disable such hints for modules that use CPP? I'm unsure whether there's a better solution.

**Comment by zliu41** [comment]
Hmm I guess it's not just "eta reduce" and "Redundant X", but most if not all hints may fire incorrectly with the presence of CPP.

**Comment by ndmitchell** [comment]
With CPP, you can break anything in any ways. But GHC warnings have the same properties too - I don't think HLint is any worse than that. I added a note into the README 1ffea9c1fcb779fb968f8d70d756b3225c97d08b so it doesn't come as a surprise, but I think it is indeed every hint, and just shows how dangerous CPP is.
